src/python.cpp

```diff
@@ -489,58 +629,57 @@ std::vector<std::string> py_class_names(PyObject* object) {
   // start adding class names
   std::vector<std::string> classNames;
 
+  {
+  PyErrorScopeGuard g_;
+
   // add the bases to the R class attribute
   Py_ssize_t len = PyTuple_Size(classes);
   for (Py_ssize_t i = 0; i < len; i++) {
     PyObject* base = PyTuple_GetItem(classes, i); // borrowed
     classNames.push_back(as_r_class(base));
   }
+  }
 
-  // return constructed class names
-  return classNames;
-
-}
-
-// wrap a PyObject
-PyObjectRef py_ref(PyObject* object,
-                   bool convert,
-                   const std::string& extraClass = "")
-{
+  // add python.builtin.object if we don't already have it
+  if (classNames.empty() || classNames.back() != "python.builtin.object") {
+    classNames.push_back("python.builtin.object");
+  }
 
-  // wrap
-  PyObjectRef ref(object, convert);
+  // if it's an iterator, include python.builtin.iterator, before python.builtin.object
+  if(PyIter_Check(object))
+    classNames.insert(classNames.end() - 1, "python.builtin.iterator");
 
-  // class attribute
-  std::vector<std::string> attrClass;
+  return eval_call(r_func_py_filter_classes, Rcpp::wrap(classNames));
```

Comment:
`Rcpp::wrap()` returns a raw `SEXP` which needs to be protected.

Reply:
Will do. Just for my own understanding, does `Rcpp::wrap()` keep the object protected until `eval_call()` returns? (btw, eval_call() also protects args that come in).

Reply:
I don't think `Rcpp::wrap()` makes any guarantees about whether or not the object it returns is protected; in general, if you're receiving an object from `Rcpp::wrap()`, you should protect it to be safe.